cpp/pagerank_module/algorithm/pagerank.cpp

```diff
@@ -0,0 +1,227 @@
+  /// because they represent position in the underlying std::vector.
+  void AddAdjacentPair(T left_node_id, T right_node_id, bool undirected = false) {
+    list_[left_node_id].push_back(right_node_id);
+    if (undirected) {
+      list_[right_node_id].push_back(left_node_id);
+    }
+  }
+
+  /// Be careful and don't call AddAdjacentPair while you have reference to this
+  /// vector because the referenced vector could be resized (moved) which means
+  /// that the reference is going to become invalid.
+  ///
+  /// @return A reference to std::vector of adjecent node ids.
+  const auto &GetAdjacentNodes(T node_id) const { return list_[node_id]; }
+
+ private:
+  std::vector<std::vector<T>> list_;
+};
+
+/// Calculates optimal borders for dividing edges in number_of_threads
+/// consecutive partitions (blocks) such that the maximal block size is minimal
+/// For example: if number_of_edges = 10 and number_of_threads = 3:
+/// optimal borders = {0, 3, 6, 10} so obtained blocks of edges
+/// are [0, 3>, [3, 6> and [6, 10>.
+///
+/// @param graph -- graph
+/// @param number_of_threads -- number of threads
+std::vector<std::uint64_t> CalculateOptimalBorders(const PageRankGraph &graph, const std::uint64_t number_of_threads) {
+  std::vector<std::uint64_t> borders;
+
+  if (number_of_threads == 0) {
+    throw std::runtime_error("Number of threads can't be zero (0)!");
+  }
+
+  for (std::uint64_t border_index = 0; border_index <= number_of_threads; border_index++) {
+    borders.push_back(border_index * graph.GetEdgeCount() / number_of_threads);
+  }
+  return borders;
+}
+
+/// Calculating PageRank block related to [lo, hi> interval of edges
+/// Graph edges are ordered by target node ids so target nodes of an interval
+/// of edges also form an interval of nodes.
+/// Required PageRank values of nodes from this interval will be sent
+/// as vector of values using new_rank_promise.
+///
+/// @param graph -- graph
+/// @param old_rank -- rank of nodes after previous iterations
+/// @param damping_factor -- damping factor
+/// @param lo -- left bound of interval of edges
+/// @param hi -- right bound of interval of edges
+/// @param new_rank_promise -- used for sending information about calculated new
+/// PageRank block
+void ThreadPageRankIteration(const PageRankGraph &graph, const std::vector<double> &old_rank, const std::uint64_t lo,
+                             const std::uint64_t hi, std::promise<std::vector<double>> new_rank_promise) {
+  std::vector<double> new_rank(graph.GetNodeCount(), 0);
+  // Calculate sums of PR(page)/C(page) scores for the entire block (from lo to hi edges).
+  for (std::size_t edge_id = lo; edge_id < hi; edge_id++) {
+    const auto [source, target] = graph.GetOrderedEdges()[edge_id];
+    // Add the score of target node to the sum.
+    new_rank[source] += old_rank[target] / graph.GetOutDegree(target);
+  }
+  new_rank_promise.set_value(new_rank);
```

Comment:
```suggestion
  new_rank_promise.set_value(std::move(new_rank));
```

Reply:
Maybe the last comment, but valuable one :smile: